constraint-solver/src/solver/backtracking.rs

```diff
@@ -0,0 +1,159 @@
+use itertools::Itertools;
+use powdr_number::{FieldElement, LargeInt};
+
+use crate::quadratic_symbolic_expression::RangeConstraintProvider;
+use crate::range_constraint::RangeConstraint;
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+use super::{Error, Solver};
+
+/// The maximum number of possible assignments to try when backtracking.
+const MAX_BACKTRACKING_WIDTH: u64 = 1 << 12;
+
+pub struct Backtracker<'a, T: FieldElement, V> {
+    solver: &'a Solver<T, V>,
+}
+
+impl<'a, T: FieldElement, V> Backtracker<'a, T, V> {
+    pub fn new(solver: &'a Solver<T, V>) -> Self {
+        Backtracker { solver }
+    }
+}
+
+impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display + Debug> Backtracker<'_, T, V> {
+    /// Returns a map of variable assignments that are unique in the constraint system.
+    /// Returns an error a contradiction was found.
+    pub fn get_unique_assignments(&self) -> Result<BTreeMap<V, T>, Error> {
+        log::debug!("Starting backtracking with maximum width {MAX_BACKTRACKING_WIDTH}");
+        let variable_sets = self.get_brute_force_candidates();
+
+        log::debug!(
+            "Found {} sets of variables with few possible assignments. Checking each set...",
+            variable_sets.len()
+        );
+
+        let unique_assignments = variable_sets
+            .iter()
+            .map(|assignment_candidates| self.find_unique_assignment(assignment_candidates))
+            // Might error out of a contradiction was found.
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            // Might return None if the assignment is not unique.
+            .flatten()
+            .collect::<Vec<_>>();
+
+        log::debug!(
+            "{} variable sets with unique assignments found",
+            unique_assignments.len()
+        );
+
+        let mut result = BTreeMap::new();
+        for (variable, value) in unique_assignments.iter().flatten() {
+            if let Some(old_value) = result.insert(variable.clone(), *value) {
+                if old_value != *value {
+                    // Two assignments contradict each other.
```

Comment:
Can this happen?

Reply:
I think yes, if you have overlapping variable sets, e.g. $\\{v_1, v_2, v_3\\}$ and $\\{v_3, v_4, v_5\\}$ might come up with different assignments for $v_3$.

Reply:
Can you add a test? Should actually be not more than
```
x + 7y + 10z = 7;
3x + 2y + z = 3;
```
plus the boolean constraints.